Telegraf startup error reporting improvement

vlastahajek opened the pull request:
In startup phase where there is not logging set yet, reports errors to OS event source (event viewer on windows). Helps debugging #4235 

- [X] Signed [CLA](https://influxdata.com/community/cla/).
- [ ] Associated README.md updated.
- [ ] Has appropriate unit tests.

vlastahajek commented:
This is helpful in the case Telegraf runs as a Windows service and error is reported before config with valid logging settings is successfully loaded.

1. Install Telegraf as a Windows service, with invalid config path
2. Start the Telegraf service, it fails.
3. How can you find out why why it failed? AFAIK stderr is not available for Windows services

danielnelson commented:
What about when a logfile is set in the agent config?  Are we still missing the startup log messages?

vlastahajek commented:
Do you mean to set log file path in [the first config inicialization](https://github.com/bonitoo-io/telegraf/blob/3c0214fb938df11c8f246821796edd0de4272fb7/cmd/telegraf/telegraf.go#L112)?

That should work. The question is what would be good place where to store startup error log.

Windows event log seem to be more suitable solution,IMHO.

glinton commented:
>Windows event log seem to be more suitable solution,IMHO.

Agreed. I'm not by any means a windows admin, but back in my windows days I would look to the event log first. It was only after using linux that i had a concept of logfiles. This pr worked well for me when i tested it yesterday.

danielnelson commented:
I think the ideal behavior here is to either use the logfile, stderr, or the event log, but probably not a mixture and it should be controllable by the admin.  What I'm gathering from this code is that it is probably much easier to support the event log than I had imagined. 
 https://github.com/influxdata/telegraf/issues/1840

vlastahajek commented:
👍 That's the idea I wasn't courageous enough to suggest.
But what about debug messages in such case? Windows Event log doesn't have such category. 
Maybe skips this for service? My idea behind that is that in case someone installs this as a service it is already tested in console mode and debug message may not be necessary.

danielnelson commented:
Yeah let's skip debug, seems like it wouldn't be appropriate to write to the Event Log that much.

vlastahajek commented:
And what about making this configurable, so it is not a breaking change. Don't use service logging unless it's turned on in config.  Sth like:
``` toml
[agent]
  interval = "2s"
  debug = "false"
  logFile = ""
  #useWindowsEventLog = true
```

It would behave as it is now in this PR. That means emergency logging before config is loaded.  Log to the Event log just in case there is an error returned from the agent start, than switch to the log defined in the config (stay with the service log  if configured so).

Another benefit of having service logger from the start is that the initial message (`Starting Telegraf `) would finally appear.

danielnelson commented:
Let's add an option like:
```
[agent]
  ## Can be one of "file" or "event_log" (windows only).
  log_handler = "file"
```

Someday, I'd like to add "syslog" too, so this will allow for that.

It occurs to me that we will need a bit more refactoring in order to send everything to the configured log location, which I don't think we should do here.  We could imagine an error like "config file not found" though, which would always occur before the config file is read.  For these cases, I think it would be fine to write to the event log in Windows.

vlastahajek commented:
Ok, so, what is the next step for this PR? 
a) Let it as it is now, as temporary fallback, to solve critical start-up problems on Windows, before logging will be enhanced
b) Should I design API for logging described  [above](https://github.com/influxdata/telegraf/pull/4291#issuecomment-515595003)?

danielnelson commented:
Let's try event log initially on Windows but switch to the `log_handler` option as early as possible.  If the `log_handler` is not set, we default to "file" and the existing `logfile` option controls if we write to stderr or open a file.

Also, since we have our current option `logfile` maybe we should name the new option `loghandler` with no underscore?  Open to better names here.

vlastahajek commented:
I've added logging to Windows Event log. However it required a bit decomposition to have logging handlers for different logging targets . As always decomposition is very subjective, so it may be not your point of view.

The name of config parameter I've chosen is `log_target`, because its values mean where log message will written. 

Name of value of event log target is `eventlog` (not snake case), to have the same form as `stdout` or `stderr`.  (I've added also `stdout` cause to have also this option, but it may be useless, not sure). 

Registering `eventlog` handler is done in the main code cause it should be available only in service mode and required parameter is event logger, which is available from service object.

There is also [new variable](https://github.com/bonitoo-io/telegraf/blob/4a7aee01881da435532851cb8654eb776db7abbf/logger/logger.go#L113) keeping track what is actually set as a log output, because log package doesn't provide a getter. It serves two purposes:
 - close previous writer if re-set
- have possibility to test what is actually set

danielnelson commented on logger/logger.go:
```diff
@@ -77,19 +122,27 @@ func newLogWriter(config LogConfig) io.Writer {
-	log.SetOutput(telegrafLog)
-	return telegrafLog
+func init() {
+	tlc := &telegrafLogCreator{}
```
This is okay though I do feel like it is a little bit overkill.  Everything is happening within the package, and all we really need is to pass over the eventlog logger and then we could initialize, without a second factory.

vlastahajek replied:
It's handy when there is a handler which is not compilable on other platforms. I first though that event log would be that case, but it turned out that kardianos lib is nicely interfaced and decouples platform dependencies. It even looks the syslog could be already solved by the `eventlog` handler, because using that lib.  If it would be that case factories are really not needed. 
Should I remove that?

danielnelson replied:
I'd probably drop it for now, we can always add back later and in fact we have some other logging stuff coming along which might change the calculus.  I'll leave it to you to decide what you think is most understandable.

vlastahajek replied:
I decided to keep that. Sure, it is a bit over-engineered, but it allows smoothly dynamically adding `eventlog` provider without 'ifs'

vlastahajek commented on docs/CONFIGURATION.md:
```diff
@@ -141,6 +141,10 @@ The agent table configures Telegraf and the defaults used across all plugins.
   Log only error level messages.
 
+- **log_target**:
+  Log target - `file`, `stderr`, `stdout` or `eventlog` (Windows only). 
```
If we are adding new option which controls the log target, wouldn't be nice to let it be the main controller of log target and have the other option (logfile) as a additional log config values param?

danielnelson replied:
I guess this would be good.  In a sense we would be deprecating `logfile="" = stderr` which was never great.

vlastahajek replied:
How to properly inform about deprecation? For sure, I will add deprecation to comment of the logfile param to configuration code and the doc. I saw that some other plugins also print warnings.
So, if config will not contain `logTarget` and it will find `logfile=""` it should print warning about deprecation and info about new option?

danielnelson replied:
Let's not formally deprecate it yet, we can do this later.

vlastahajek commented on docs/CONFIGURATION.md:
```diff
@@ -141,6 +141,10 @@ The agent table configures Telegraf and the defaults used across all plugins.
 - **quiet**:
   Log only error level messages.
 
+- **log_target**:
```
Ok

vlastahajek commented on cmd/telegraf/telegraf.go:
```diff
@@ -430,13 +437,24 @@ func main() {
 				(*svcConfig).Arguments = append((*svcConfig).Arguments, "--config-directory", *fConfigDirectory)
 			}
+			//set servicename to service cmd line, to have a custom name after relaunch as a service
+			(*svcConfig).Arguments = append((*svcConfig).Arguments, "--service-name", *fServiceName)
```
Service name is mandatory. `fServiceName` has default value `telegraf`. To check if it was change it requires keeping track of original value and comparing. What is a drawback of having `--service-name` param always?

vlastahajek replied:
We can use just `svcConfig.Arguments`, I was blindly copying similar code above.

danielnelson replied:
If we have a default then no problem, just was concerned it could be null

vlastahajek commented on logger/event_logger_test.go:
```diff
@@ -0,0 +1,91 @@
+}
+
+func TestEventLog(t *testing.T) {
+	prepareLogger(t)
```
Well, I did this the same way as in `logger_tests.go` (all platforms). All tests are run always and create (delete) log files.
Actually, It creates just 4 messages in event log, where there is a lot other useless messages from other applications, which is not bad. 
I see much more offending some other telegraf tests which creates some exec stubs in the temp folder and each run several firewall windows  appears to allow connection :(

danielnelson replied:
Certainly not our worst transgression, but still add the `-short` skip.

vlastahajek commented on cmd/telegraf/telegraf.go:
```diff
@@ -152,6 +153,9 @@ func runAgent(ctx context.Context,
 	// Setup default logging. This may need to change after reading the config
 	// file, but we can configure it to use our logger implementation now.
+	if !initLoggingSet {
+		logger.SetupLogging(logger.LogConfig{})
```
Removing this would wipe the `evenlog` setting created in the service section. 
Other option is to have a global  variable for initial default config and modify it just for the service mode. The var will than be used here.

danielnelson replied:
I don't follow, wouldn't this only be called if the event log couldn't be initialized?  In that case this has already be called in main.

vlastahajek replied:
[In this part](https://github.com/bonitoo-io/telegraf/blob/vh-startup-error-reporting/cmd/telegraf/telegraf.go#L450) ,
``` go
           } else {
			winlogger, err := s.Logger(nil)
			if err == nil {
				//When in service mode, register eventlog target andd setup default logging to eventlog
				logger.RegisterEventLogger(winlogger)
				logger.SetupLogging(logger.LogConfig{LogTarget: logger.LogTargetEventlog})
				initLoggingSet = true
			}
			err = s.Run()

			if err != nil {
				log.Println("E! " + err.Error())
			}
		}
```
I'm setting default config option to `eventlog` **only** when telegraf runs on windows as a service and it runs before [`runAgent`](https://github.com/bonitoo-io/telegraf/blob/vh-startup-error-reporting/cmd/telegraf/telegraf.go#L157),
``` go
func runAgent(ctx context.Context,
	inputFilters []string,
	outputFilters []string,
) error {
	// Setup default logging. This may need to change after reading the config
	// file, but we can configure it to use our logger implementation now.
	if !initLoggingSet {
		logger.SetupLogging(logger.LogConfig{})
	}
	log.Printf("I! Starting Telegraf %s", version)
``` 
which runs always.
If there is no check for pre-set default logging, this part will re-set previously set eventlog to stderr

danielnelson replied:
I'm not saying remove the check, I think we can remove all of this change.  If we remove it then Windows will be fine and wouldn't non-Windows also be okay because we are now calling SetupLogging earlier in main?

vlastahajek replied:
I overlooked setting up default logging in main. So, sure, this part in `runAgent` so useless.

vlastahajek commented on logger/logger.go:
```diff
@@ -2,33 +2,33 @@ package logger
 
 import (
 	"errors"
+	"github.com/influxdata/telegraf/internal/rotate"
```
Ok. need to figure out how to set this in IDE (IntelliJ Idea/Goland)

danielnelson commented on logger/logger.go:
```diff
@@ -77,19 +122,27 @@ func newLogWriter(config LogConfig) io.Writer {
-	} else {
-		writer = os.Stderr
+	if closer, isCloser := actualLogger.(io.Closer); isCloser {
+		closer.Close()
```
Not sure if its an issue, but can you double check that we don't close stderr?  Also, we may not want to close the logfile as it could mean we move back to stderr logging on reload.

vlastahajek replied:
Good point. But actual check was needed in the `telegrafLog.Close`, which is responsible for closing internal log.
